internal/provider/resource_ssh_key.go

```diff
@@ -5,183 +5,309 @@ import (
-		return diag.FromErr(err)
+	modelName, user := retrieveSSHKeyInfoFromID(plan.ID.ValueString(), &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	if err = d.Set("model", result.ModelName); err != nil {
-		return diag.FromErr(err)
+	modelInfo, err := s.client.Models.GetModelByName(modelName)
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to resolve model UUID, got error: %s", err))
+		return
 	}
 
-	if err = d.Set("payload", result.Payload); err != nil {
-		return diag.FromErr(err)
+	result, err := s.client.SSHKeys.ReadSSHKey(&juju.ReadSSHKeyInput{
+		ModelName: modelInfo.Name,
+		ModelUUID: modelInfo.UUID,
+		User:      user,
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read ssh key, got error: %s", err))
+		return
 	}
+	tflog.Trace(ctx, fmt.Sprintf("read ssh key resource %q", plan.ID.ValueString()))
 
-	d.SetId(fmt.Sprintf("sshkey:%s:%s", modelName, user))
+	plan.ModelName = types.StringValue(result.ModelName)
+	plan.Payload = types.StringValue(result.Payload)
 
-	return diag.Diagnostics{}
+	// Set the plan onto the Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
 }
 
-func sshKeyUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	var diags diag.Diagnostics
+func (s *sshKeyResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	// Prevent panic if the provider has not been configured.
+	if s.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "ssh_key", "update")
+		return
+	}
+
+	var plan, state sshKeyResourceModel
 
-	client := meta.(*juju.Client)
+	// Get the Terraform state from the request into the state model
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	// Read Terraform configuration from the request into the plan model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
-	if !d.HasChange("payload") {
-		return diags
+	// Return early if nothing has changed
+	if plan.Payload.Equal(state.Payload) && plan.ModelName.Equal(state.ModelName) {
+		return
 	}
 
-	// any change in the payload has to be considered as a new key
-	modelName := d.Get("model").(string)
-	modelUUID, err := client.Models.ResolveModelUUID(modelName)
+	modelName := state.ModelName.ValueString()
+	modelUUID, err := s.client.Models.ResolveModelUUID(modelName)
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to resolve model UUID, got error: %s", err))
+		return
 	}
 
-	user := utils.GetUserFromSSHKey(d.Get("payload").(string))
+	payload := state.Payload.ValueString()
+	user := utils.GetUserFromSSHKey(payload)
 	if user == "" {
-		return diag.Errorf("malformed SSH key, user not found")
+		resp.Diagnostics.AddError("Client Error", "malformed SSH key, user not found")
```

Comment:
Let's have a different error message, if we get a malformed ssh key from req.State - that feels like a provider bug, not something the user can fix.